README.md
# cpp-skeleton
## Features

#### Automatic implementations at _.cpp_ files creation giving the matching _.hpp files_.
**The file has to start with an upper-case letter, making it a class .cpp file.**

![oops something went wrong, try to check on https://raw.githubusercontent.com/ldedier/cpp-skeleton/master/medias/cppGif.gif](https://raw.githubusercontent.com/ldedier/cpp-skeleton/master/medias/cppGif.gif)

## Extension Settings

#### Specify these configurations in *User Settings* :

```ts
{
  "cpp-skeleton.headerCommandId": string, //a vscode command ID to execute before the skeleton creation
  "cpp-skeleton.generateSettersAndGetters": boolean, // automatic implementations of getters and setters
}
```

### Commands

You can execute these commands by acceding to the execute console (cmd + shift + P)

#### Add missing methods

![oops something went wrong, try to check on https://raw.githubusercontent.com/ldedier/cpp-skeleton/master/medias/cppGif2.gif](https://raw.githubusercontent.com/ldedier/cpp-skeleton/master/medias/cppGif2.gif)

Checks for methods present in the .hpp file that are not implemented and add them to the current .cpp file

 #### Clear as new skeleton

Clear your .cpp file and rebuild the skeleton from the matching .hpp file

License
----
MIT


CHANGELOG.md
# Change Log

All notable changes to the "cpp-skeleton" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/extension.ts
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   extension.ts                                       :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: ldedier <ldedier@student.42.fr>            +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2019/07/12 20:04:21 by ldedier           #+#    #+#             */
/*   Updated: 2019/07/12 20:42:39 by ldedier          ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import CppWatcher from './CppWatcher';
import CppImplementer from './CppImplementer';

const watcher : CppWatcher = new CppWatcher();

export function activate(context: vscode.ExtensionContext) {

	let disposable = vscode.commands.registerCommand('cpp-skeleton.addMissingMethods', CppImplementer.implementsMissingMethods);
	let disposable2 = vscode.commands.registerCommand('cpp-skeleton.clearAsNewSkeleton', CppImplementer.clearAsNewSkeleton);

	context.subscriptions.push(disposable);
	context.subscriptions.push(disposable2);
	
}

export function deactivate() {}


src/CppWatcher.ts
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   CppWatcher.ts                                      :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: ldedier <ldedier@student.42.fr>            +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2019/07/12 20:04:18 by ldedier           #+#    #+#             */
/*   Updated: 2019/07/12 20:04:35 by ldedier          ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

import * as vscode from 'vscode';
import CppImplementer from './CppImplementer';

class CppWatcher {

	public fileSystemWatcher: vscode.FileSystemWatcher;

	constructor()
	{
		this.fileSystemWatcher = vscode.workspace.createFileSystemWatcher("**/[A-Z]{[A-Z],[a-z]}*.cpp",
			false, true, true);
		this.fileSystemWatcher.onDidCreate(this.onCreateCpp);
	}

	private onCreateCpp = (uri: vscode.Uri) => {
		if (vscode.window.activeTextEditor
			&& vscode.window.activeTextEditor.document.uri.path == uri.path)
		{
			let editor = vscode.window.activeTextEditor as vscode.TextEditor;
			const matchinHppFileName = vscode.window.activeTextEditor.document.fileName.replace(/\.cpp/, ".hpp");
			vscode.workspace.openTextDocument(matchinHppFileName).then(
			(hppDoc) => {
				const cppImplementer = new CppImplementer(editor, hppDoc);
				cppImplementer.fillSkeleton();
			},
			(err) => {
				vscode.window.showWarningMessage(err.message);
			})
		}
	}

	public implementMissingMethods()
	{

	}
}
export default CppWatcher;

src/Implementation.ts
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   Implementation.ts                                  :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: ldedier <ldedier@student.42.fr>            +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2019/07/12 20:04:26 by ldedier           #+#    #+#             */
/*   Updated: 2019/07/12 21:33:43 by ldedier          ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

class CppImplementation 
{
	prototype: string;
	body: string;
	methodName: string;
	parameters: string;

	constructor(prototype: string, body: string, methodName: string, parameters: string)
	{
		this.prototype = prototype;
		this.body = body;
		this.methodName = methodName;
		this.parameters = parameters;
	}

	public implements(): string
	{
		return (this.prototype + "\n{\n" + this.body + "\n}");
	}
}
export default CppImplementation;

src/HppClassDecomposer.ts
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   HppClassDecomposer.ts                              :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: ldedier <ldedier@student.42.fr>            +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2019/07/12 20:04:23 by ldedier           #+#    #+#             */
/*   Updated: 2019/07/12 20:04:33 by ldedier          ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

import * as vscode from 'vscode'

class HppClassDecomposer
{
	public text: string;
	public top: string;
	public innerClass: string;
	public bottom: string;
	public publicPart: string;
	public privatePart: string;
	public success: boolean;

	constructor(document: vscode.TextDocument)
	{
		this.text = document.getText();
		
		const regex : RegExp =  new RegExp(/^([^]*)class[^]*{([^]*public:([^]*)private:([^]*))};([^]*)$/, 'g');
		const array: RegExpExecArray | null = regex.exec(document.getText());
		if (array == null)
		{
			this.top = "";
			this.innerClass = "";
			this.publicPart = "";
			this.privatePart = "";
			this.bottom = "";
			this.text = "";
			this.success = false;
		}
		else
		{
			this.text = array[0];
			this.top = array[1];
			this.innerClass = array[2];
			this.publicPart = array[3];
			this.privatePart = array[4];
			this.bottom = array[5];
			this.success = true;
		}
	}

}



export default HppClassDecomposer;

src/CppImplementer.ts
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   CppImplementer.ts                                  :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: ldedier <ldedier@student.42.fr>            +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2019/07/12 20:04:15 by ldedier           #+#    #+#             */
/*   Updated: 2019/07/12 20:14:43 by ldedier          ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

import * as vscode from 'vscode';
import HppClassDecomposer from './HppClassDecomposer';
import CppImplementation from './Implementation';

class CppImplementer
{

	public command: string;
	public editor: vscode.TextEditor;
	public hppDoc: vscode.TextDocument;
	public generateSettersAndGetters: boolean;

	constructor(editor: vscode.TextEditor, hppDoc: vscode.TextDocument)
	{
		this.editor = editor;
		this.hppDoc = hppDoc;
		this.command = (vscode.workspace.getConfiguration().get('cpp-skeleton.headerCommandId') as string).trim();
		this.generateSettersAndGetters = vscode.workspace.getConfiguration().get('cpp-skeleton.generateSettersAndGetters') as boolean;
	}

	private processFillSkeletonFromHpp()
	{
		var ImplementationsArray: CppImplementation[] | null;
		ImplementationsArray = this.GetImplementationsArray();
		if (ImplementationsArray != null)
		{
			this.editor.edit( edit => {
				const pos: vscode.Position = this.editor.selection.active;
				edit.insert(pos,
					[
						"#include \""+ this.hppDoc.fileName.split("/").pop() as string +"\"",
						(<CppImplementation[]> ImplementationsArray).map(
							implementation => implementation.implements()).join("\n\n")
					].join("\n\n"));
			});
		}
		else
		{
			vscode.window.showErrorMessage(this.hppDoc.fileName + " is not a well formatted hpp class file.");
		}
	}

	private  computeImplementation(regexArray: RegExpExecArray, className: string, innerClass: boolean): CppImplementation
	{
		var type = regexArray[4];
		var rhs = regexArray[10];
		const stc = regexArray[2];
		const methodName = regexArray[11];
		const args = regexArray[12];

		if (rhs.charAt(0) == '(')
		{
			rhs = type + rhs;
			type = "";
		}

		var prototype: string;
		var body = "\t";
		// res = (stc ? "/*\n** static method\n*/\n" : "");
		let arr: RegExpMatchArray | null;
		if (!stc && innerClass && this.generateSettersAndGetters &&  (arr = methodName.match(new RegExp(/.*get([A-Za-z_0-9\-~].*)/g, ""))))
			body = "\treturn (this->_" + arr[1].charAt(0).toLowerCase() + arr[1].slice(1) + ");";
		else if (!stc && innerClass && this.generateSettersAndGetters && (arr = methodName.match(new RegExp(/.*set([A-Za-z_0-9\-~].*)/g, ""))))
			body = "\tthis->_" + arr[1].charAt(0).toLowerCase() + arr[1].slice(1) + " = " + args.split(" ").slice(-1)[0] + ";";
		const classNamePrefix = (innerClass ? className + "::" : "");
		if (type)
			prototype = type + "\t" + classNamePrefix + rhs;
		else
			prototype = classNamePrefix + rhs;
		return (new CppImplementation(prototype, body, methodName, args));
	}

	private GetImplementationsArray(): CppImplementation[] | null
	{
		const decomposer : HppClassDecomposer = new HppClassDecomposer(this.hppDoc);
		if (decomposer.success)
		{
			
			const Functionregex = RegExp(/^(\t| )*(static)?(\t| )*(([A-Za-z_:\-~]*)((\t| )*(&|\*))?)(\t| )*(([A-Za-z_0-9<=\+\/\-~]*)\((.*)\).*);$/, 'gm');
			let array: RegExpExecArray | null;
			var implementations: CppImplementation[] = [];
			const className : string = (this.editor.document.fileName.split("/").pop() as string).split(".")[0];
			while ((array = Functionregex.exec(decomposer.innerClass)))
			{
				let implementation = this.computeImplementation(array, className, true);
				implementations.push(implementation);
			}
			while ((array = Functionregex.exec(decomposer.bottom)))
			{
				let implementation = this.computeImplementation(array, className, false);
				implementations.push(implementation);
			}
			return (implementations);
		}
		else
			return (null);
	}

	public  fillSkeleton()
	{
		if (this.command && this.command.length > 0)
		{
			vscode.commands.executeCommand(this.command).then(() => {
				vscode.window.showTextDocument(this.editor.document, 1, false).then(() => {
					this.processFillSkeletonFromHpp();
				})
			}, (err)=> {
				vscode.window.showErrorMessage(err.message);
			})
		}
		else
			this.processFillSkeletonFromHpp();
	}

	private processImplementMissingMethods()
	{
		const text = this.editor.document.getText();
		
		var implementationsArray: CppImplementation[] | null;
		var missingImplementations: CppImplementation[] = [];
		if ((implementationsArray = this.GetImplementationsArray()))
		{
			console.log('text', text);

			for (let i = 0; i < implementationsArray.length; i++) {
				const implementation = implementationsArray[i];
				if (text.indexOf(implementation.prototype) == -1)
					missingImplementations.push(implementationsArray[i])
			}
			if (missingImplementations.length > 0)
			{
				this.editor.edit( edit => {
					const pos: vscode.Position = this.editor.document.positionAt(this.editor.document.getText().length);
					// this.editor.selection.active = pos;
					edit.insert(pos,
						[	"",
							(<CppImplementation[]> missingImplementations).map(elt => elt.implements()).join("\n\n")
						].join("\n\n"));
				});
			}
		}
		else
			vscode.window.showErrorMessage(this.hppDoc.fileName + " is not a well formatted hpp class file.");
	}

	public static implementsMissingMethods()
	{
		if (vscode.window.activeTextEditor
			&& vscode.window.activeTextEditor.document.fileName.split(".").slice(-1)[0] == "cpp")
		{
			let editor = vscode.window.activeTextEditor as vscode.TextEditor;
			const matchinHppFileName = vscode.window.activeTextEditor.document.fileName.replace(/\.cpp/, ".hpp");
			vscode.workspace.openTextDocument(matchinHppFileName).then(
			(doc) => {
					const implementer : CppImplementer = new CppImplementer(editor, doc); 
					implementer.processImplementMissingMethods();
			},
			(err) => {
				vscode.window.showErrorMessage(err.message);
			})
		}
		else
			vscode.window.showErrorMessage("this command shall be executed while editing a .cpp file")
	}

	public static clearAsNewSkeleton()
	{
		if (vscode.window.activeTextEditor
			&& vscode.window.activeTextEditor.document.fileName.split(".").slice(-1)[0] == "cpp")
		{
			let editor = vscode.window.activeTextEditor as vscode.TextEditor;
			const matchinHppFileName = vscode.window.activeTextEditor.document.fileName.replace(/\.cpp/, ".hpp");
			vscode.workspace.openTextDocument(matchinHppFileName).then(
			(doc) => {
				var firstLine = editor.document.lineAt(0);
				var lastLine = editor.document.lineAt(editor.document.lineCount - 1);
				var textRange = new vscode.Range(0, 
								 firstLine.range.start.character, 
								 editor.document.lineCount - 1, 
								 lastLine.range.end.character);
				editor.edit(edit =>
				{
					edit.delete(textRange);
				}).then(() => {
					const implementer : CppImplementer = new CppImplementer(editor, doc); 
					implementer.fillSkeleton();
					})
			},
			(err) => {
				vscode.window.showErrorMessage(err.message);
			})
		}
		else
			vscode.window.showErrorMessage("this command shall be executed while editing a .cpp file")
	}
}
export default CppImplementer;